### PR TITLE
Instruction codes

### PR DESCRIPTION
**Pull request by fantypants**
- New Tests for all instruction codes started
- Instruction Codes in Progress, with latest updates to corresponding codes
- Code Parser File tidied and Organized
- Executor Module Organized

### CODE REVIEW

File: lib/execution/executor.ex [hunk 1]
```diff
@@ -530,28 +440,226 @@ defmodule WaspVM.Executor do
...
     {[b, a], stack} = Stack.pop_multiple(vm.stack)
-    j2 = b - (32 * Integer.floor_div(b, 32))
-
-    answer = rotl(a, j2)
 
-    {frame, Map.put(vm, :stack, Stack.push(stack, answer))}
+    if b == 0 do
+      {:error, :undefined}
+    else
+      {frame, Map.put(vm, :stack, Stack.push(stack, a - (b*trunc(a/b))))}
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: This isn't very readable, lets split this into a few steps, perhaps by piping? Something like

```elixir
c = 
  a
  |> Kernel./(b)
  |> trunc()
  |> Kernel.*(b)

res = a - c

```

File: lib/execution/executor.ex [hunk 2]
```diff
@@ -530,28 +440,226 @@ defmodule WaspVM.Executor do
...
-  defp exec_inst({frame, vm}, :i32_rotr) do
+  defp exec_inst({frame, vm}, :i64_rem_u) do
     {[b, a], stack} = Stack.pop_multiple(vm.stack)
-    j2 = Integer.mod(b, 32)
 
-    answer = rotr(a, j2)
+    if b == 0 do
+      {:error, :undefined}
+    else
+      {frame, Map.put(vm, :stack, Stack.push(stack, a - (b*trunc(a/b))))}
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Same as above

File: lib/execution/executor.ex [hunk 3]
```diff
@@ -530,28 +440,226 @@ defmodule WaspVM.Executor do
...
+  defp exec_inst({frame, vm}, :f64_floor) do
+    {[a], stack} = Stack.pop_multiple(vm.stack)
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, Float.floor(a)))}
+  end
+
+  defp exec_inst({frame, vm}, :f32_neg) do
+    {a, stack} = Stack.pop(vm.stack)
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, float_point_op(a*-1)))}
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Add spacing between operators

```elixir
a * -1
```

File: lib/execution/executor.ex [hunk 4]
```diff
@@ -530,28 +440,226 @@ defmodule WaspVM.Executor do
...
+  end
+
+  defp exec_inst({frame, vm}, :f64_neg) do
+    {a, stack} = Stack.pop(vm.stack)
+
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, float_point_op(a*-1)))}
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Same as above

File: lib/execution/executor.ex [hunk 5]
```diff
@@ -530,28 +440,226 @@ defmodule WaspVM.Executor do
...
+    {[a, b], stack} = Stack.pop_multiple(vm.stack)
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, float_point_op(copysign(a, b))))}
+  end
+
+
+  defp exec_inst({frame, vm}, :f32_abs) do
+    {a, stack} = Stack.pop(vm.stack)
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, float_point_op(abs(a))))}
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Lets try not to nest operations, do this instead

```elixir
a
|> abs()
|> float_point_op()
```

File: lib/execution/executor.ex [hunk 6]
```diff
@@ -530,28 +440,226 @@ defmodule WaspVM.Executor do
...
+  end
+
+  defp exec_inst({frame, vm}, :f64_abs) do
+    {[a], stack} = Stack.pop_multiple(vm.stack)
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, float_point_op(abs(a))))}
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Same as above

File: lib/execution/executor.ex [hunk 7]
```diff
@@ -530,28 +440,226 @@ defmodule WaspVM.Executor do
...
+  defp exec_inst({frame, vm}, :f64_sqrt) do
+    {[a], stack} = Stack.pop(vm.stack)
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, :math.sqrt(a)))}
+  end
+
+  defp exec_inst({frame, vm}, :f32_div) do
+    {[b, a], stack} = Stack.pop_multiple(vm.stack)
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, WaspVM.Executor.float_point_op(a / b)))}
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: You should remove the `WaspVM.Executor`

File: lib/execution/executor.ex [hunk 8]
```diff
@@ -580,343 +684,390 @@ defmodule WaspVM.Executor do
...
+
+    {frame, Map.put(vm, :stack, Stack.push(stack, bsl(a, b)))}
+  end
+
+
+
+  defp exec_inst({frame, vm}, :i32_shr_u) do
+    {[b, a], stack} = Stack.pop_multiple(vm.stack)
+
+      j2 = Integer.mod(32, b)
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Indentation needs fixing

File: lib/execution/executor.ex [hunk 9]
```diff
@@ -945,11 +1100,57 @@ defmodule WaspVM.Executor do
...
+    b_truth =
+      to_string(b)
+      |> String.codepoints
+      |> Enum.any?(&(&1 == "-"))
+
+    if a_truth == true && b_truth == true || a_truth == false && b_truth == false  do
+      a
+    else
+      if a_truth == true && b_truth == false || a_truth == false && b_truth == true do
+        b*-1 |> IO.inspect
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Space between operators + remove inspect

File: lib/execution/executor.ex [hunk 10]
```diff
@@ -945,11 +1100,57 @@ defmodule WaspVM.Executor do
...
+      end
+    end
   end
+
+
+  defp check_value([0, b, c, d]) when b and c and d !== 0, do: 1
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Nice! This is a good optimization

File: lib/execution/executor.ex [hunk 11]
```diff
@@ -945,11 +1100,57 @@ defmodule WaspVM.Executor do
     |> Enum.count()
   end
 
-  defp rotl(number, shift) do
-    number <<< shift ||| number >>> (32 - shift)
+  defp rotl(number, shift), do: (number <<< shift) ||| (number >>> (0x1F &&& (32 + ~~~(shift + 1)))) &&& ~~~(0xFFFFFFFF <<< shift)
+  defp rotr(number, shift), do: (number >>> shift) ||| (number <<< (0x1F &&& (32 + ~~~(-shift + 1)))) &&& ~~~(0xFFFFFFFF <<< -shift)
+  def float_point_op(number) do
+    D.set_context(%D.Context{D.get_context | precision: 6})
```
Comment (alexdovzhanyn) [inline lib/execution/executor.ex]: Why use decimals here instead of regular floats?

File: lib/execution/executor.ex (hunk 7 above)
Comment (fantypants) [reply lib/execution/executor.ex]: Resolved

File: lib/execution/executor.ex (hunk 9 above)
Comment (fantypants) [reply lib/execution/executor.ex]: Resolved

File: lib/execution/executor.ex (hunk 11 above)
Comment (fantypants) [reply lib/execution/executor.ex]: The specs designated the floats had to be to a certain decimal. I worked off the Web Assembly Binary tool for a reference, regular floats I'm not 100% would satisfy it. Main thing is i wanted consistent output i.e not 0.0, and then 1.134567.

**Pull request merged**